File: src/main/java/com/github/novotnyr/idea/git/CheckoutNewBranchAction.java
package com.github.novotnyr.idea.git;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator;
import com.intellij.openapi.project.Project;
import git4idea.GitUtil;
import git4idea.branch.GitBrancher;
import git4idea.repo.GitRepository;
import git4idea.repo.GitRepositoryManager;
import git4idea.validators.GitNewBranchNameValidator;
import org.jetbrains.annotations.CalledInAwt;
import org.jetbrains.annotations.CalledInBackground;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public class CheckoutNewBranchAction extends AnAction {
    public CheckoutNewBranchAction() {
        super("Checkout New Branch...", "Checkout and switch to a new branch", null);
    }

    @Override
    public void actionPerformed(AnActionEvent anActionEvent) {
        Project project = anActionEvent.getProject();

        List<SelectedModule> selectedModules = SelectedModule.manyOf(anActionEvent);
        if (selectedModules.isEmpty()) {
            return;
        }

        switchBranch(project, selectedModules);
    }

    private void switchBranch(Project project, Collection<SelectedModule> selectedModule) {
        GitBrancher gitBrancher = GitBrancher.getInstance(project);
        GitRepositoryManager repositoryManager = GitUtil.getRepositoryManager(project);

        ProgressManager progressManager = ProgressManager.getInstance();
        Task.Backgroundable task = new Task.Backgroundable(project, "Retrieving Git branches", true) {
            private ConcurrentMap<GitRepository, String> branchMapping = new ConcurrentHashMap<>();

            @Override
            @CalledInBackground
            public void run(@NotNull ProgressIndicator indicator) {
                for (SelectedModule module : selectedModule) {
                    GitRepository repo = repositoryManager.getRepositoryForFile(module.getFile());
                    if (repo == null) {
                        return;
                    }
                    this.branchMapping.put(repo, GitUtil.HEAD);
                }
            }

            @Override
            @CalledInAwt
            public void onSuccess() {
                List<GitRepository> repositories = new ArrayList<>(this.branchMapping.keySet());
                GitNewBranchOptions options = getNewBranchNameFromUser(project, repositories);
                if (options != null) {
                    if (options.shouldCheckout()) {
                        gitBrancher.checkoutNewBranch(options.getName(), repositories);
                    } else {
                        gitBrancher.createBranch(options.getName(), this.branchMapping);
                    }
                }
            }
        };
        var progressIndicator = new BackgroundableProcessIndicator(task);
        progressManager.runProcessWithProgressAsynchronously(task, progressIndicator);

    }

    private GitNewBranchOptions getNewBranchNameFromUser(Project project, List<GitRepository> repositories) {
        String dialogTitle = "Create New Branch";
        return getNewBranchNameFromUser(project, repositories, dialogTitle, "");
    }

    /**
     * <p>Shows a message dialog to enter the name of new branch.</p>
     * <p>Optionally allows to not checkout this branch, and just create it.</p>
     *
     * <p>Mirrors {@code git4idea.branch.GitBranchUtil#getNewBranchNameFromUser(com.intellij.openapi.project.Project, java.util.Collection, java.lang.String, java.lang.String)}</p>
     *
     * @return the name of the new branch and whether it should be checked out, or {@code null} if user has cancelled the dialog.
     */
    @Nullable
    public static GitNewBranchOptions getNewBranchNameFromUser(@NotNull Project project,
                                                               @NotNull Collection<GitRepository> repositories,
                                                               @NotNull String dialogTitle,
                                                               @Nullable String initialName) {
        return new GitNewBranchDialog(project, dialogTitle, initialName, GitNewBranchNameValidator.newInstance(repositories)).showAndGetOptions();
    }
}

File: src/main/java/com/github/novotnyr/idea/git/CheckoutNewBranchAction.kt
package com.github.novotnyr.idea.git

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.impl.BackgroundableProcessIndicator
import com.intellij.openapi.project.Project
import git4idea.branch.GitBrancher
import git4idea.repo.GitRepository
import git4idea.validators.GitNewBranchNameValidator

class CheckoutNewBranchAction : AnAction("Checkout New Branch...", "Checkout and switch to a new branch", null) {
    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project ?: return

        val selectedModules = SelectedModule.manyOf(e)
        if (selectedModules.isEmpty()) {
            return
        }
        switchBranch(project, selectedModules)
    }

    private fun switchBranch(project: Project, selectedModule: Collection<SelectedModule>) {
        val gitBrancher = GitBrancher.getInstance(project)
        val task = RetrieveGitBranchesTask(project, selectedModule) { branchMapping ->
            val repositories = ArrayList(branchMapping.keys)
            val options = getNewBranchNameFromUser(project, repositories) ?: return@RetrieveGitBranchesTask
            if (options.shouldCheckout()) {
                gitBrancher.checkoutNewBranch(options.name, repositories)
            } else {
                gitBrancher.createBranch(options.name, branchMapping)
            }
        }
        ProgressManager.getInstance().runProcessWithProgressAsynchronously(task, BackgroundableProcessIndicator(task))
    }

    private fun getNewBranchNameFromUser(project: Project, repositories: List<GitRepository>): GitNewBranchOptions? {
        val dialogTitle = "Create New Branch"
        return getNewBranchNameFromUser(project, repositories, dialogTitle, "")
    }

    /**
     *
     * Shows a message dialog to enter the name of new branch.
     * Optionally allows to not checkout this branch, and just create it.
     * Mirrors `git4idea.branch.GitBranchUtil#getNewBranchNameFromUser(com.intellij.openapi.project.Project, java.util.Collection, java.lang.String, java.lang.String)`
     *
     * @return the name of the new branch and whether it should be checked out, or `null` if user has cancelled the dialog.
     */
    fun getNewBranchNameFromUser(
        project: Project,
        repositories: Collection<GitRepository>,
        dialogTitle: String,
        initialName: String?
    ): GitNewBranchOptions? {
        return GitNewBranchDialog(
            project,
            dialogTitle,
            initialName,
            GitNewBranchNameValidator.newInstance(repositories)
        ).showAndGetOptions()
    }
}